contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,188 @@
+    bytes32 public immutable currentBlockchainID;
+    bytes32 public immutable sourceBlockchainID;
+    address public immutable sourceContractAddress;
+    // We will not mint the first `tokenReserve` tokens sent to this subnet.
+    // This should be constructed to match the initial token supply of this subnet.
+    // This will mean we will not mint tokens until the source contact is collateralized.
+    uint256 public tokenReserve;
+
+    // Used for sending an receiving Teleporter messages.
+    ITeleporterMessenger public immutable teleporterMessenger;
+
+    // TODO we probably want to add the original token supply from this chain to the constructor.
+    constructor(
+        address teleporterMessengerAddress,
+        bytes32 sourceBlockchainID_,
+        address sourceContractAddress_,
+        uint256 tokenReserve_
+    ) {
+        currentBlockchainID = WarpMessenger(WARP_PRECOMPILE_ADDRESS)
+            .getBlockchainID();
+
+        require(
+            teleporterMessengerAddress != address(0),
+            "Invalid Teleporter Messenger Address"
+        );
+        teleporterMessenger = ITeleporterMessenger(teleporterMessengerAddress);
+
+        require(sourceBlockchainID_ != bytes32(0), "Invalid Source Chain ID");
+        require(
+            sourceBlockchainID_ != currentBlockchainID,
+            "Cannot Bridge With Same Blockchain"
+        );
+        sourceBlockchainID = sourceBlockchainID_;
+
+        require(
+            sourceContractAddress_ != address(0),
+            "Invalid Source Contract Address"
+        );
+        sourceContractAddress = sourceContractAddress_;
+
+        tokenReserve = tokenReserve_;
+    }
+
+    /**
+     * @dev See {ITeleporterReceiver-receiveTeleporterMessage}.
+     *
+     * Receives a Teleporter message and routes to the appropriate internal function call.
+     */
+    function receiveTeleporterMessage(
+        bytes32 senderBlockchainID,
+        address senderAddress,
+        bytes calldata message
+    ) external nonReentrant {
+        // Only allow the Teleporter messenger to deliver messages.
+        require(
+            msg.sender == address(teleporterMessenger),
+            "Unauthorized teleporter contract"
+        );
+
+        // Only allow messages from the source chain.
+        require(
+            senderBlockchainID == sourceBlockchainID,
+            "Invalid Source Chain"
+        );
+
+        // Only allow the partner contract to send messages.
+        require(senderAddress == sourceContractAddress, "Unauthorized Sender");
+
+        (address recipient, uint256 amount) = abi.decode(
+            message,
+            (address, uint256)
+        );
+        require(recipient != address(0), "Invalid Recipient Address");
+
+        uint256 adjustedAmount = amount;
+        if (tokenReserve > 0) {
+            if (amount > tokenReserve) {
+                emit CollateralAdded({amount: tokenReserve, remaining: 0});
```

Comment:
I'm not quitesure what we should emit for `amount` in this case. Say `tokenReserve=10` on startup initialization. The effective collateral locked on the source side is therefore -10. We then bridge 100 tokens over, and mint 90. There's now 100 tokens as collateral on the source, and 100 total native tokens on the destination subnet. Should we emit the amount as 100 (the amount of collateral) or 110 ( 100 - (-10) )? Either way, I don't think `amount: tokenReserve` is correct. Please let me know if I'm off the mark here.

Reply:
Perhaps we should emit `CollateralAdded` from the source contract. We'd probably need to add `tokenReserve` to the `NativeTokenSource` constructor in that case.

Reply:
My idea for the event emitted here was to only emit `CollateralAdded` for the first 10 tokens in your example. Otherwise we just emit `MintNativeTokens`.

We discussed which contract to have control over the collateralization, and we came to the conclusion of using the destination contract, I think because if we end up in the ERC20-to-native bridge example and we want to worry about adjusting the number of decimals, then only the destination contract still only needs to know the number of coins minted at genesis. However the source contract decides to convert decimals will be completely up to the source contract. 

Not sure if I explained that well, but we can chat about it offline if you'd like.